pkg/cmd/roachtest/operations/changefeeds/utils.go

```diff
@@ -0,0 +1,407 @@
+	return helpers.ExecuteQuery(ctx, func(rowValues []string) error {
+		options := make(map[string]string)
+		err := json.Unmarshal([]byte(rowValues[1]), &options)
+		if err != nil {
+			return err
+		}
+		jobID := rowValues[0]
+		jobIDToDetails[jobID].payload = &jobspb.ChangefeedDetails{
+			Opts: options, SinkURI: rowValues[2],
+		}
+		return err
+	}, conn, stmt)
+}
+
+// fetchAllChangefeedJobsWithStates fetches all changefeed jobs with the given states.
+func fetchAllChangefeedJobsWithStates(
+	ctx context.Context, conn *gosql.DB, states ...jobs.State,
+) ([]*jobDetails, error) {
+	// Build SQL query to fetch jobs with specified states.
+	statement := buildQueryToGetCFJobsWithStates(states...)
+	details := make([]*jobDetails, 0)
+
+	// Execute query and populate job details.
+	err := helpers.ExecuteQuery(ctx, func(rowValues []string) error {
+		details = append(details, &jobDetails{
+			jobID: rowValues[0],
+			state: jobs.State(rowValues[1]),
+		})
+		return nil
+	}, conn, statement)
+
+	return details, err
+}
+
+// buildQueryToGetCFJobsWithStates constructs the SQL query to fetch changefeed jobs.
+func buildQueryToGetCFJobsWithStates(states ...jobs.State) string {
+	whereClause := ""
+	// If states are provided, append them to the WHERE clause.
+	if len(states) > 0 {
+		stateStrs := make([]string, 0)
+		for _, s := range states {
+			stateStrs = append(stateStrs, string(s))
+		}
+		whereClause = fmt.Sprintf(" WHERE status in ('%s')", strings.Join(stateStrs, "', '"))
+	}
+
+	// Construct and return the final SQL statement.
+	return fmt.Sprintf("SELECT job_id, status FROM [SHOW CHANGEFEED JOBS]%s;", whereClause)
+}
+
+// getJobIDState retrieves the current state of a specific job by job ID.
+func getJobIDState(ctx context.Context, conn *gosql.DB, jobID string) (jobs.State, error) {
+	// Construct SQL query to get job state by job ID.
+	statement := fmt.Sprintf("SELECT status FROM [SHOW CHANGEFEED JOBS] WHERE job_id='%s';", jobID)
+
+	var state jobs.State
+	// Execute query and extract the state value.
+	err := helpers.ExecuteQuery(ctx, func(rowValues []string) error {
+		state = jobs.State(rowValues[0])
+		return nil
+	}, conn, statement)
+
+	return state, err
+}
+
+// changeStateForChangefeed executes the specified action (e.g., PAUSE, CANCEL) on a randomly
+// selected changefeed job from the list of jobs in currentStates.
+// It confirms the success by checking if the job transitions to one of the expectedFinalStates.
+// If no existing jobs are found it attempts to create a changefeed.
+// It uses the checkAndCreateChangefeed function to determine if a new changefeed can be created.
+func changeStateOrCreateChangefeed(
+	ctx context.Context,
+	o operation.Operation,
+	conn *gosql.DB,
+	action string,
+	currentStates, expectedFinalStates []jobs.State,
+) error {
+	// Fetch all changefeed jobs with the given current states.
+	cfJobs, err := fetchAllChangefeedJobsWithStates(ctx, conn, currentStates...)
+	if err != nil {
+		o.Fatal(err)
```

Comment:
why does this error cause an o.Fatal but others return err?

Reply:
This was a miss. Thanks for pointing out.